src/Fable.Transforms/Global/Babel.fs

```diff
@@ -219,32 +219,236 @@ module PrinterExtensions =
+            match this with
+            | Declaration(s) -> printer.Print(s)
+            | IfStatement(s) -> printer.Print(s)
+            | TryStatement(s) -> printer.Print(s)
+            | ForStatement(s) -> printer.Print(s)
+            | BreakStatement(s) -> printer.Print(s)
+            | WhileStatement(s) -> printer.Print(s)
+            | ThrowStatement(s) -> printer.Print(s)
+            | BlockStatement(s) -> printer.Print(s)
+            | ReturnStatement(s) -> printer.Print(s)
+            | SwitchStatement(s) -> printer.Print(s)
+            | LabeledStatement(s) -> printer.Print(s)
+            | DebuggerStatement(s) -> printer.Print(s)
+            | ContinueStatement(s) -> printer.Print(s)
+            | ExpressionStatement(s) -> printer.Print(s)
+
 /// Note that declarations are considered statements; this is because declarations can appear in any statement context.
-type Declaration = inherit Statement
+type Declaration =
+    | ClassDeclaration of ClassDeclaration
+    | VariableDeclaration of VariableDeclaration
+    | FunctionDeclaration of FunctionDeclaration
+    | InterfaceDeclaration of InterfaceDeclaration
+
+    interface IPrintable with
+        member this.Print(printer) =
+            match this with
+            | ClassDeclaration(d) -> printer.Print(d)
+            | VariableDeclaration(d) -> printer.Print(d)
+            | FunctionDeclaration(d) -> printer.Print(d)
+            | InterfaceDeclaration(d) -> printer.Print(d)
+
 /// A module import or export declaration.
-type ModuleDeclaration = inherit Node
+type ModuleDeclaration =
+    | ImportDeclaration of ImportDeclaration
+    | ExportAllDeclaration of ExportAllDeclaration
+    | ExportNamedReferences of ExportNamedReferences
+    | ExportNamedDeclaration of ExportNamedDeclaration
+    | PrivateModuleDeclaration of PrivateModuleDeclaration
+    | ExportDefaultDeclaration of ExportDefaultDeclaration
+
+    interface IPrintable with
+        member this.Print(printer) =
+            this.Print(printer)
+
+    member this.Print(printer: Printer) =
+            match this with
+            | ImportDeclaration(d) -> printer.Print(d)
+            | ExportAllDeclaration(d) -> printer.Print(d)
+            | ExportNamedReferences(d) -> printer.Print(d)
+            | ExportNamedDeclaration(d) -> printer.Print(d)
+            | PrivateModuleDeclaration(d) -> printer.Print(d)
+            | ExportDefaultDeclaration(d) -> printer.Print(d)
 
 /// Not in Babel specs
-type EmitExpression(value, args, ?loc) =
-    member _.Value: string = value
-    member _.Args: Expression array = args
-    interface Expression with
-        member _.Print(printer) =
-            printer.AddLocation(loc)
+type EmitExpression =
+    { Value: string
+      Args: Expression array
+      Loc: SourceLocation option }
+
+    static member AsExpr(value, args, ?loc): Expression =
+        { Value = value
+          Args = args
+          Loc = loc }
+        |> EmitExpression
```

Comment:
Sorry @dbrattli, I think you mentioned this somewhere but, what was the purpose of using records instead of just union case fields? Would it be simpler to do something like:

```fsharp
type Expression =
    | EmitExpression of value: string * args: Expression array * loc: SourceLocation option
    ...
    static member NewEmitExpression(value, args, ?loc) =
        EmitExpression(value, args, loc)
```

Would it make the code simpler to do this in general and use only standalone types when they're needed?

Reply:
This is basically just a matter of style. Records is very similar to named unions fields (except it's easier to extract values from records using dot-notation). The question is really about the extra type-level. This is great of attaching convenience methods such as `Print` and `Create`. But if you like a more compact form then I can try to reduce it. There will however be cases like `Expression` that will need to split the printing into several different functions, so not so compact anymore. But if we can more the print to a separate file (it's just another transformation AST -> string) then that could be good. I'll see what I can do.

Reply:
Ok, I've tried to fix this. But the Babel AST is quite self referencing so it's not possible to do this for many of the types and they need to be types, e.g `Identifier`. `EmitExpression` is easy since it's not part of Babel. But I think it looks better so please tell me what you think. I would like to get back to Python sometime soon ;)

Reply:
Ok, think I'm done now. There's quite a bit of commented code in Babel.fs that is not relevant anymore wrt the new style etc. Should we delete it?